### PR TITLE
WIP: dimod refactor

### PR DESCRIPTION
**Pull request by arcondello**

### CODE REVIEW

**Comment by arcondello** [comment]
I consider template_response to be in final form (for this commit) if you guys want to take a look there. I'll update you as each file gets updated. @bellert @orenshk

File: dimod/responses/template_response.py [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+>>> list(response.samples())
+[{'a': -1, 'b': -1}, {'a': 1, 'b': 1}, {'a': -1, 'b': 1}, {'a': 1, 'b': -1}]
+
+Note that the samples are returned in order of increasing energy. The energies
+can also be queried
+
+>>> list(response.energies())
+[-1.5, -0.5, -0.5, 2.5]
+
+Finally, if there is a data associated with any of the samples, it can accessed through
```
Comment (orenshk) [inline dimod/responses/template_response.py]: Missing 'be', "...it can *be* accessed"

File: dimod/responses/template_response.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+            [({0: 1}, -1), ({0: -1}, 1)]
+            >>> list(response.items(data=True))
+            [({0: 1}, -1, {'n': 1}), ({0: -1}, 1, {'n': 5})]
+
+        Note:
+            Depreciation Warning: This method of access is being depreciated.
+            it can be replaced by `response.data(keys=['sample', 'energy'])`
+
+        """
+        warnings.warn("'items' will be depreciated in dimod 1.0.0", DeprecationWarning)
```
Comment (orenshk) [inline dimod/responses/template_response.py]: `''.join('depreciated'.split('i'))`

File: dimod/responses/template_response.py [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+
+        Examples:
+            >>> response = dimod.TemplateResponse()
+            >>> response.add_samples_from([{0: -1}, {0: 1}], [1, -1], dataval='test')
+            >>> list(response.data())
+            [{'dataval': 'test', 'energy': -1, 'num_occurences': 1, 'sample': {0: 1}},
+             {'dataval': 'test', 'energy': 1, 'num_occurences': 1, 'sample': {0: -1}}]
+
+        """
+        stop = object()  # signals that the two iterables are different lengths
```
Comment (orenshk) [inline dimod/responses/template_response.py]: Why not `None`?

File: dimod/responses/template_response.py (hunk 2 above)
Comment (bellert) [reply dimod/responses/template_response.py]: deprecated, not depreciated

File: dimod/responses/template_response.py (hunk 3 above)
Comment (arcondello) [reply dimod/responses/template_response.py]: object has a unique ID, but None does not.

**Pull request merged**